crates/wasi-common/src/snapshots/wasi_snapshot_preview1.rs

```diff
@@ -0,0 +1,1448 @@
+use crate::entry::{Descriptor, Entry};
+use crate::sandboxed_tty_writer::SandboxedTTYWriter;
+use crate::wasi::wasi_snapshot_preview1::WasiSnapshotPreview1;
+use crate::wasi::{types, AsBytes, Errno, Result};
+use crate::WasiCtx;
+use crate::{clock, fd, path, poll};
+use log::{debug, error, trace};
+use std::cell::Ref;
+use std::convert::TryInto;
+use std::io::{self, Read, Seek, SeekFrom, Write};
+use std::ops::DerefMut;
+use wiggle_runtime::{GuestBorrows, GuestPtr};
+
+impl<'a> WasiSnapshotPreview1 for WasiCtx {
+    fn args_get<'b>(
```

Comment:
@sunfishcode Note that we don't have `unsafe`'s in methods' signatures. I was somewhat worried about this since we're still dealing with file descriptors and other OS handles, but perhaps the entry points to WASI syscalls are OK being safe especially given that WASI fd is now a handle type? What do you think?

Reply:
Crazy idea that I haven't thought all the way through yet: Would it work if we made `Fd` work like `std::fs::File`?
 - Make `Fd::from_raw` unsafe? , similar to [`from_raw_fd`](https://doc.rust-lang.org/std/fs/struct.File.html#method.from_raw_fd) being unsafe.
 - Make `Fd` *not* implement `Copy`. Change `Fd` arguments to `&Fd`.
 - Make ` Fd` implement `Drop` and have it automatically close. This might be hard though. Is there a way in Rust to make `drop` private, so that you have to pass it back to the context?

Reply:
On further consideration, the `Drop` part makes this unworkable. I'm going to investigate other approaches.

Reply:
I like the idea but yeah, `Drop` will not work like this, unless we made the context object static/global. Having said that your suggestion aligns with my mental model of `Fd` 100%, so perhaps we could come up with something close to it but not necessarily one-to-one.

Reply:
Ok, thinking this through more carefully, maybe we can do this without the `Drop` part. So make `handle` types not implement `Copy`, pass `handle` arguments by reference rather than by value, and make the raw constructor `unsafe`. That's not the same as `std::fs::File`, but it still has some nice properties. We just need to be careful when calling the raw constructor, but there aren't many places where we need to do that outside of the generated wrappers. Does that sound feasible?

If you'd like to defer this to a later PR, that's fine too.

Reply:
Sounds good, and yeah, if we could defer to a subsequent PR, that would be perfect IMHO. There's already a substantial amount of tweaks I have to apply and I'm worried more and more will slip away. Is that OK?